### PR TITLE
add new exit code for SYSTEM_PERIODIC_REMOVE

### LINKED ISSUE #7906: Special Error for SYSTEM_PERIODIC_REMOVE
@thongonary and I were looking at 

https://cmst2.web.cern.ch/cmst2/unified//condorlogs/pdmvserv_task_HIN-pPb816Spring16GS-00049__v1_T_170601_155641_5744/99303/HIN-pPb816Spring16GS-00049_0/1047/vocms0304_1047339/Job_1047339/condor.12807.0.log

and the error code in wmstats is 99303, while 

    <a n="MachineAttrGLIDEIN_CMSSite0"><s>T2_CH_CERN</s></a>
    <a n="Proc"><i>0</i></a>
    <a n="TriggerEventTypeName"><s>ULOG_JOB_ABORTED</s></a>
    <a n="EventTime"><s>2017-06-03T01:38:56</s></a>
    <a n="EnteredCurrentStatus"><i>1496446736</i></a>
    <a n="ExtDESIRED_Sites"><s>T1_DE_KIT,T1_ES_PIC,T1_FR_CCIN2P3,T1_RU_JINR,T1_UK_RAL,T2_AT_Vienna,T2_BE_IIHE,T2_BE_UCL,T2_BR_SPRACE,T2_CH_CERN,T2_CH_CERN_HLT,T2_CH_CSCS,T2_CN_Beijing,T2_DE_DESY,T2_DE_RWTH,T2_ES_CIEMAT,T2_ES_IFCA,T2_FR_CCIN2P3,T2_FR_GRIF_IRFU,T2_FR_GRIF_LLR,T2_FR_IPHC,T2_GR_Ioannina,T2_HU_Budapest,T2_IT_Bari,T2_IT_Legnaro,T2_IT_Rome,T2_KR_KNU,T2_PL_Swierk,T2_PT_NCG_Lisbon,T2_RU_JINR,T2_TR_METU,T2_TW_NCHC,T2_UK_London_Brunel,T2_UK_London_IC,T2_UK_SGrid_Bristol,T2_UK_SGrid_RALPP,T2_US_MIT,T3_IN_TIFRCloud</s></a>
    <a n="QDate"><i>1496410842</i></a>
    <a n="TriggerEventTypeNumber"><i>9</i></a>
    <a n="WMAgent_JobID"><i>1047339</i></a>
    <a n="MyType"><s>JobAdInformationEvent</s></a>
    <a n="Cluster"><i>12807</i></a>
    <a n="JobStatus"><i>3</i></a>
    <a n="Reason"><s>The system macro SYSTEM_PERIODIC_REMOVE expression '( ( NumJobStarts &gt; 9 ) || ( NumShadowStarts &gt; 19 ) || ( DiskUsage &gt; 27000000 ) )' evaluated to TRUE</s></a>

is there a way to not have 99303, but something that has more meaning ?

### PR DESCRIPTION
**Pull request by ticoann**
fixes #7906

### CODE REVIEW

**Comment by amaltaro** [comment]
@ticoann Seangchan, Diego D. applied a small change to my VMs (vocms0123 and vocms0125) in order to get the RemoveReason class ad. I still have to check/test that, then I'll get back here.

**Comment by amaltaro** [comment]
Better later than never :)
@ticoann I finally managed to test the condor changes made by Diego D. and it seems to be working just fine. So, when a job fails, we might look for `Reason` in the condor.log file. I understand it will be there whenever a job is killed via condor (by user or by glideinwms itself).

Let me know if you want me to make a new PR instead. Please find an example condor.log here:
http://amaltaro.web.cern.ch/amaltaro/forWMCore/PR_8100/condor.8003.40.log

**Comment by ticoann** [comment]
@amaltaro Alan, Sorry, I missed your comment. Now we can search for the reason but do we have to map specific reason to specific exit code?

What is the relation between

via condor_rm (by user cmst1) and SYSTEM_PERIODIC_REMOVE?

**Comment by amaltaro** [comment]
I think the SYSTEM_PERIODIC_REMOVE is more general and it's all about glideinWMS job removal. Either because the job hit one of the constraints, or exceeded resources allocated, or as we could see, was removed by a person.

I don't think it's going to be straight forward to map different reasons to different exit code, not as a start at least. The most important IMO, is that we propagate the real reason that a job was removed from the pool. Then later we can make a fine-grained error report.

You have lots of stuff on your back, so I'm taking this feature from here. If needed I can bug Diego and get further help.

**Comment by ticoann** [comment]
Alan you have more things to do. 😄 
My impression was @vlimant Jean-Roch wants to have different exit code at least for this specific case. Anyway, I modified the patch to get the Reason part and only filters 2 cases (SYSTEM_PERIODIC_REMOVE, via condor_rm). It is not the best solution we can refine it later.

File: src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py [hunk 1]
```diff
@@ -289,7 +291,15 @@ def complete(self, jobs):
                             logOutput += '\n\n'
                             for exMsg in listWMExceptionStr(condorFilePath):
                                 logOutput += "\n\n%s\n" % exMsg
-                    condorReport.addError("NoJobReport", 99303, "NoJobReport", logOutput)
+
+                            if condorFile == condorLog:
+                                for condorReason in BasicAlgos.findMagicStr(condorFilePath, '<a n="Reason">'):
```
Comment (amaltaro) [inline src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py]: You probably need to add `logOutput += '\n\n'` after this for loop. Just to separate the reason from further log lines.
In addition to that, I expect to see `Reason` only once, but this implementation is good.

File: src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py [hunk 2]
```diff
@@ -289,7 +291,15 @@ def complete(self, jobs):
...
+                                    logOutput += condorReason
+                                    if "SYSTEM_PERIODIC_REMOVE" in condorReason or "via condor_rm" in condorReason:
+                                        exitCode = 99400
+                                        exitType = "RemoveByGLIDEIN"
```
Comment (amaltaro) [inline src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py]: And here, first comment: how about renaming it to `RemovedByGLIDEIN`?

2nd comment: how about we put an else and set the same exitType but make exitCode = 99401, for instance. Later on we can learn what are the cases for 99401 and apply further filter.

**Comment by vlimant** [comment]
At first order, having an exit code different than 99303 for when SYSTEM_PERIODIC_REMOVE is reached (so that we can distinguish htcondor issue from something bad happening with the job, from the exit code)

Of course having something finer grained would be nice, but not really realistic since the condition that failed in htcondor could change with no warning (or do we have control over the expression ?). Best would be that htcondor gives a well identified error code back and wmagent propagate this through ( @bbockelm ?)

**Comment by ticoann** [comment]
Thanks Alan, I updated the patch. @vlimant, Is the exit code OK? I randomly picked. If there is a better exit code we should change.

**Comment by amaltaro** [comment]
Looks good, Seangchan. If you feel like adding it to the upcoming agent, go for it and we can have another round of tests.

**Comment by ticoann** [comment]
I will run the quick test tonight before I create the new tag for agent. Thanks Alan.

**Comment by ticoann** [comment]
At least it didn't break anything I didn't try to abort jobs though. Merging it and will do the further test.

**Comment by amaltaro** [comment]
I expected to see 99400 exit code for
amaltaro_StepChain_DupOutMod_Agent116_Validation_CC7_170918_155824_9635

but it has just the standard 99303 error. I'll investigate it further tomorrow.

**Comment by amaltaro** [comment]
If you insist :-)
`cmst1@vocms0224:/data/srv/wmagent/current $ vim install/wmagent/JobArchiver/logDir/a/amaltaro_StepChain_DupOutMod_Agent116_Validation_CC7_170918_155824_9635/JobCluster_1/Job_1543.tar.bz2 `

thanks!

**Comment by ticoann** [comment]
Thanks Alan.

**Comment by ticoann** [comment]
The problem seems to be race condition.
When GlideIn killed the job condor log is still being written. so there is no Reason found. so it reports 99303 but after framework job report created it finishes the condor log.

Here is the condor log with the line number
```
450 <c>
451     <a n="Proc"><i>2</i></a>
452     <a n="Size"><i>1772288</i></a>
453     <a n="EventTime"><s>2017-09-18T17:14:46</s></a>
454     <a n="EnteredCurrentStatus"><i>1505744063</i></a>
455     <a n="WMAgent_JobID"><i>1543</i></a>
456     <a n="QDate"><i>1505743998</i></a>
457     <a n="TriggerEventTypeName"><s>ULOG_IMAGE_SIZE</s></a>
458     <a n="MemoryUsage"><i>807</i></a>
459     <a n="DESIRED_Sites"><s>T1_US_FNAL,T2_CH_CERN</s></a>
460     <a n="JobStartDate"><i>1505744063</i></a>
461     <a n="ResidentSetSize"><i>826200</i></a>
462     <a n="MyType"><s>JobAdInformationEvent</s></a>
463     <a n="TriggerEventTypeNumber"><i>6</i></a>
464     <a n="Cluster"><i>19315</i></a>
465     <a n="JobStatus"><i>2</i></a>
466     <a n="Subproc"><i>0</i></a>
467     <a n="EventTypeNumber"><i>28</i></a>
468     <a n="MachineAttrGLIDEIN_CMSSite0"><s>T1_US_FNAL</s></a>
469     <a n="ExtDESIRED_Sites"><s>T1_US_FNAL,T2_CH_CERN</s></a>
470 </c>
471 <c>
472     <a n="Proc"><i>2</i></a>
473     <a n="TerminatedAndRequeued"><b v="f"/></a>
474     <a n="Checkpointed"><b v="f"/></a>
475     <a n="EventTime"><s>2017-09-18T17:15:40</s></a>
476     <a n="TerminatedNormally"><b v="f"/></a>
477     <a n="ReceivedBytes"><r>2.996012000000000E+06</r></a>
478     <a n="RunLocalUsage"><s>Usr 0 00:00:00, Sys 0 00:00:00</s></a>
479     <a n="SentBytes"><r>0.0</r></a>
480     <a n="RunRemoteUsage"><s>Usr 0 00:57:14, Sys 0 00:01:13</s></a>
481     <a n="MyType"><s>JobEvictedEvent</s></a>
482     <a n="Cluster"><i>19315</i></a>
483     <a n="Subproc"><i>0</i></a>
484     <a n="EventTypeNumber"><i>4</i></a>
485 </c>
486 <c>
487     <a n="Proc"><i>2</i></a>
488     <a n="TerminatedAndRequeued"><b v="f"/></a>
489     <a n="Checkpointed"><b v="f"/></a>
490     <a n="EventTime"><s>2017-09-18T17:15:40</s></a>
491     <a n="EnteredCurrentStatus"><i>1505744063</i></a>
492     <a n="WMAgent_JobID"><i>1543</i></a>
493     <a n="TerminatedNormally"><b v="f"/></a>
494     <a n="QDate"><i>1505743998</i></a>
495     <a n="TriggerEventTypeName"><s>ULOG_JOB_EVICTED</s></a>
496     <a n="ReceivedBytes"><r>2.996012000000000E+06</r></a>
497     <a n="RunLocalUsage"><s>Usr 0 00:00:00, Sys 0 00:00:00</s></a>
498     <a n="DESIRED_Sites"><s>T1_US_FNAL,T2_CH_CERN</s></a>
499     <a n="JobStartDate"><i>1505744063</i></a>
500     <a n="SentBytes"><r>0.0</r></a>
501     <a n="RunRemoteUsage"><s>Usr 0 00:57:14, Sys 0 00:01:13</s></a>
502     <a n="MyType"><s>JobAdInformationEvent</s></a>
503     <a n="TriggerEventTypeNumber"><i>4</i></a>
504     <a n="Cluster"><i>19315</i></a>
505     <a n="JobStatus"><i>2</i></a>
506     <a n="Subproc"><i>0</i></a>
507     <a n="EventTypeNumber"><i>28</i></a>
508     <a n="MachineAttrGLIDEIN_CMSSite0"><s>T1_US_FNAL</s></a>
509     <a n="ExtDESIRED_Sites"><s>T1_US_FNAL,T2_CH_CERN</s></a>
510 </c>
511 <c>
512     <a n="Reason"><s>via condor_rm (by user cmst1)</s></a>
513     <a n="MyType"><s>JobAbortedEvent</s></a>
514     <a n="Proc"><i>2</i></a>
515     <a n="Cluster"><i>19315</i></a>
516     <a n="EventTime"><s>2017-09-18T17:15:40</s></a>
517     <a n="Subproc"><i>0</i></a>
518     <a n="EventTypeNumber"><i>9</i></a>
519 </c>
```

And here is FWJR error details.
```
Could not find jobReport\nAdding end of condor.19315.2.log to error message:\n\n    <a n="EventTime"><s>2017-09-18T17:09:46</s></a>\n    <a n="EnteredCurrentStatus"><i>1505744063</i></a>\n    <a n="WMAgent_JobID"><i>1543</i></a>\n    <a n="QDate"><i>1505743998</i></a>\n    <a n="TriggerEventTypeName"><s>ULOG_IMAGE_SIZE</s></a>\n    <a n="MemoryUsage"><i>807</i></a>\n    <a n="DESIRED_Sites"><s>T1_US_FNAL,T2_CH_CERN</s></a>\n    <a n="JobStartDate"><i>1505744063</i></a>\n    <a n="ResidentSetSize"><i>826200</i></a>\n    <a n="MyType"><s>JobAdInformationEvent</s></a>\n    <a n="TriggerEventTypeNumber"><i>6</i></a>\n    <a n="Cluster"><i>19315</i></a>\n    <a n="JobStatus"><i>2</i></a>\n    <a n="Subproc"><i>0</i></a>\n    <a n="EventTypeNumber"><i>28</i></a>\n    <a n="MachineAttrGLIDEIN_CMSSite0"><s>T1_US_FNAL</s></a>\n    <a n="ExtDESIRED_Sites"><s>T1_US_FNAL,T2_CH_CERN</s></a>\n</c>\n<c>\n    <a n="Size"><i>1772288</i></a>\n    <a n="MyType"><s>JobImageSizeEvent</s></a>\n    <a n="MemoryUsage"><i>807</i></a>\n    <a n="Proc"><i>2</i></a>\n    <a n="Cluster"><i>19315</i></a>\n    <a n="EventTime"><s>2017-09-18T17:14:46</s></a>\n    <a n="Subproc"><i>0</i></a>\n    <a n="EventTypeNumber"><i>6</i></a>\n    <a n="ResidentSetSize"><i>826200</i></a>\n</c>\n<c>\n    <a n="Proc"><i>2</i></a>\n    <a n="Size"><i>1772288</i></a>\n    <a n="EventTime"><s>2017-09-18T17:14:46</s></a>\n    <a n="EnteredCurrentStatus"><i>1505744063</i></a>\n    <a n="WMAgent_JobID"><i>1543</i></a>\n    <a n="QDate"><i>1505743998</i></a>\n    <a n="TriggerEventTypeName"><s>ULOG_IMAGE_SIZE</s></a>\n    <a n="MemoryUsage"><i>807</i></a>\n    <a n="DESIRED_Sites"><s>T1_US_FNAL,T2_CH_CERN</s></a>\n    <a n="JobStartDate"><i>1505744063</i></a>\n    <a n="ResidentSetSize"><i>826200</i></a>\n    <a n="MyType"><s>JobAdInformationEvent</s></a>\n    <a n="TriggerEventTypeNumber"><i>6</i></a>\n    <a n="Cluster"><i>19315</i></a>\n    <a n="JobStatus"><i>2</i></a>\n    <a n="Subproc"><i>0</i></a>\n    <a n="EventTypeNumber"><i>28</i></a>\n    <a n="MachineAttrGLIDEIN_CMSSite0"><s>T1_US_FNAL</s></a>\n    <a n="ExtDESIRED_Sites"><s>T1_US_FNAL,T2_CH_CERN</s></a>\n</c>\n\n\n\n\n
```

**Comment by ticoann** [comment]
1. We can simply wait a few seconds before open the file(~10sec) or 
2. We can check whether file is opened by other process and wait until it is closed.

Ref.
https://stackoverflow.com/questions/11114492/check-if-a-file-is-not-open-not-used-by-other-process-in-python

Alan, What do you think?

**Comment by ticoann** [comment]
Patch is created #8178, not sure whether that will fix the problem though. If that looks OK, we can add to a new agent as well.

**Pull request merged**